File: bitcalctess/tessellator/calc.py
from sympy import divisors

def bs_equals(bs1, bs2):
	"""are bitstrings equal"""
	return bs1 == bs2

def bl_equals(bs1, bs2):
	"""are bitloops equal"""
	return len(bs1) == len(bs2) and bs2 in bs1 + bs1

def dec_to_bs(n, dec):
	"""convert decimal to binary, preserve string length"""
	b = (bin(dec))[2:]
	return (n - len(b)) * '0' + b

def bs_to_dec(bs):
	"""convert binary to decimal"""
	bs = bs[bs.find('1'):]
	return int(str(bs), 2)

def parity(bs):
	"""sum of 1s in the bitstring is even (true/false)"""
	return bs.count('1') % 2 == 0

def inverse(bs):
	"""switch 1s for 0s and vice versa"""
	i = ''
	for b in bs:
		i += str(int(b) ^ 1)
	return i

def is_inverse(bs):
	"""does the rotation class of the bitloop
	contain inverse pairs of bitstrings"""
	return bl_equals(bs, inverse(bs))

def reverse(bs):
	"""reverse order of bits"""
	return bs[::-1]

def is_reverse(bs):
	"""is reverse the same as non-reverse"""
	return bs == reverse(bs)

def is_symmetric(bs):
	"""does the rotation class of the bitloop
	contain reverse pairs of bitstrings"""
	for i in rotations(bs):
		if bs == reverse(i):
			return True
	return False

def rotate(bs, t):
	"""rotate order of bits by t bits"""
	t %= len(bs)
	r = bs[t:] + bs[:t]
	return r

def link(bs):
	"""return the link of a given bitloop"""
	l = ''
	for i in range(len(bs)):
		l += str(int(bs[i]) ^ int(bs[(i +1) % len(bs)]))
	return min_val_rot(l)

def sublink(bs):
	"""return the sublink of a given bitloop
	or an empty string if the bitloop has no sublink"""
	sl = ''
	if parity(bs):
		sl += '0'
		for i in range(len(bs) - 1):
			sl += str(int(sl[i]) ^ int(bs[i]))
	return sl if sl == '' else min_val_rot(sl)

def pattern(bs):
	"""return the pattern of a given bitloop"""
	n = len(bs)
	divs = divisors(n)
	for div in divs:
		p = bs[:div]
		if p * (n // div) == bs:
			return p

def rotations(bs):
	"""return all unique rotations of a bitstring
	that is, the rotation class of the respective bitloop"""
	rots = []
	for i in range(len(pattern(bs))):
		rots.append(rotate(bs, i))
	return rots

def min_val_rot(bs):
	"""minimum value rotation (canonical form)"""
	rots = [bs_to_dec(r) for r in rotations(bs)]
	mn = dec_to_bs(len(bs), min(rots))
	return (len(bs) - len(mn)) * '0' + mn

def root(bs):
	"""the root of the tree that a bitloop is in"""
	cycle = [bs]
	lbs = link(bs)
	while not lbs in cycle:
		cycle.append(lbs)
		lbs = link(lbs)
	return min_val_rot(lbs)

def is_root(bs):
	"""is the bitloop a tree root"""
	return bs == root(bs)

def num_roots(rt):
	"""number of roots in a cycle"""
	l = link(rt)
	i = 1
	while not bl_equals(rt, l):
		i += 1
		l = link(l)
	return i

def roots(rt, num_roots):
	"""list of roots in a chain
	input a root and num_roots(root)"""
	rts = []
	for i in range(num_roots):
		rts.append(rt)
	rt = link(rt)
	return rts

def min_val_root(roots):
	"""minimum value root given the roots in a chain"""
	for r in roots:
		r = bs_to_dec(r)
	return min(roots)

def sublink_and_root(root):
	"""the sublink of a root that is also a root"""
	s = sublink(root)
	return s if is_root(s) else inverse(s)

def num_lvl_tree(bs):
	"""number of levels in a tree"""
	l = 1
	bs = inverse(root(bs))
	while sublink(bs) != '':
		l += 1
		bs = sublink(bs)
	return l

def structure(bs):
	"""the structure of a tree"""
	s = '2'
	bs = inverse(root(bs))
	while sublink(bs) != '':
		bs = sublink(bs)
		s += '1' if is_inverse(bs) else '2'
	s += '0'
	return s

def num_bl_tree(bs):
	"""number of bitloops in a tree"""
	s = structure(bs)
	if len(s) == 2: return 1
	num_bl_lv = num_bl_tr = 1
	for lv in range(2, len(s)):
		num_bl_lv *= int(s[lv - 1])
		num_bl_tr += num_bl_lv
	return num_bl_tr

def num_bs_tree(bs):
	"""number of bitstrings in a tree"""
	s = structure(bs)
	bs = inverse(root(bs))
	num_bs_tr = len(pattern(bs))
	if len(s) == 2: return num_bs_tr
	num_bl_lv = 1
	for lv in range(2, len(s)):
		num_bl_lv *= int(s[lv - 1])
		bs = sublink(bs)
		num_bs_tr += len(pattern(bs)) * num_bl_lv
	return num_bs_tr

def num_bl_chain(bs):
	"""number of bitloops in a chain"""
	return num_roots(bs) + num_roots(bs) * num_bl_tree(bs)

def num_bs_chain(bs):
	"""number of bitstrings in a chain"""
	bs = root(bs)
	num_bs_cyc = len(pattern(bs)) * num_roots(bs)
	return num_bs_cyc + num_roots(bs) * num_bs_tree(bs)

def tree_nodes(root):
	"""returns the set of all bitloops in a tree"""
	t_ns = [root]
	sl_off_cycle = inverse(sublink_and_root(root))
	make_tree(sl_off_cycle, t_ns)
	return t_ns

def make_tree(t_root, t_ns):
	"""helper function for tree_nodes"""
	t_ns.append(t_root)
	sl = sublink(t_root)
	if sl == '':
		return
	inv = inverse(sl)
	make_tree(sl, t_ns)
	if not bl_equals(sl, inv):
		make_tree(inv, t_ns)

File: bitcalctess/calculator/cpp/calc.cpp
#include <iostream>
#include <vector>
#include "calc.h"
#include "util.h"
using namespace std;


bool bsEqual(int n, vector<unsigned char> a, vector<unsigned char> b) {
	// are bitstrings equal
    for (int i = 0; i < n; i++) {
		// compare bits
		if (getBit(a, i) ^ getBit(b, i)) { 
			return false;
		}
	}
	return true;
}


bool blEqual(int n, vector<unsigned char> a, vector<unsigned char> b) {
	// are bitloops equal
	for (int t = 0; t < n; t++) {
		if (bsEqual(n, a, b)) {
			return true;
		}
		rotate1(n, b);
	}
	return false;
}


void rotate1(int n, vector<unsigned char>& bs) {
	// rotate 1 bit around
	// save rightmost bit to wrap around
	unsigned char last_bit = (bs[(n - 1) / 8] >> ((n - 1) % 8)) & 1;
	// for each bit starting with the rightmost
	for (int i = n - 1; i > 0; i--) {
		// if left bit is set
		if (bs[(i - 1) / 8] & (1 << ((i - 1) % 8))) {
			// then set the bit
			bs[i / 8] |= (1 << (i % 8)); // 1 == 0b00000001
		// else it is clear
		} else {
			// clear the bit
			bs[i / 8] &= ~(1 << (i % 8));
		}
	}
	// if the rightmost bit is set
	if (last_bit) {
		// set the leftmost bit
		bs[0] |= (1 << 0);
	// else it is clear
	} else {
		// clear the bit
		bs[0] &= ~(1 << 0);
	}
}


//void rotateT(int n, vector<unsigned char>& array, int t);
void rotateT(int n, vector<unsigned char>& array, int t) {
	// rotate T bits around
	t = t % n;
	if (t < 0) {
		t += n;
	}
	vector<unsigned char> temp = initArray(n);
	for (int i = 0; i < n; i++) {
		int j = (i + t) % n;
		if (array[i / 8] & (1 << (i % 8))) {
			temp[j / 8] |= (1 << (j % 8));
		}
	}
	for (int i = 0; i < (n + 7) / 8; i++) {
		array[i] = temp[i];
	}
}


vector<unsigned char> minValRot(int n, vector<unsigned char> bs) {
	int min = 1 << n;
	for (int i = 0; i < n; i++) {
		rotate1(n, bs);
		int dec = bsToDec(n, bs);
		if (dec < min) {
			min = dec;
		}
	}
	return decToBs(n, min);
}


bool parity(int n, vector<unsigned char> bs) {
	// is bitstring "even" or "odd"
	// returns 0 (false) if there are an even number of 1s
	// returns 1 (true) if there are an odd number of 1s
	bool p = 0;
	for (int i = 0; i < n; i++) {
		bool b = getBit(bs, i);
		p = p ^ b;
	}
	return p;
}


vector<unsigned char> inverse(int n, vector<unsigned char> bs) {
	// flip 1s and 0s in a bitstring
	for (int i = 0; i < n; i++) {
		xorBit(bs, i);
	}
	return bs;
}


//bool isInverse(int n, vector<unsigned char> bs);
bool isInverse(int n, vector<unsigned char> bs) {
	// is a bitloop equal to its inverse
	vector<unsigned char> inv = inverse(n, bs);
	bool b = blEqual(n, bs, inv);
	return b;
}


//vector<unsigned char> reverse(int n, vector<unsigned char> bs);
vector<unsigned char> reverse(int n, vector<unsigned char> bs) {
	// returns bits in reverse order
	vector<unsigned char> rev = initArray(n);
	for (int i = 0; i < n; i++) {
		if (getBit(bs, n - 1 - i)) {
			setBit(rev, i);
		}
	}
	return rev;
}


//bool isReverse(int n, vector<unsigned char> bs);
bool isReverse(int n, vector<unsigned char> bs) {
	// is a bitstring equal to its reverse
	vector<unsigned char> rev = copyArray(n, bs);
	rev = reverse(n, rev);
	bool b = bsEqual(n, bs, rev);
	return b;
}


//bool isSymmetric(int n, vector<unsigned char> bs);
bool isSymmetric(int n, vector<unsigned char> bs) {
	// is a bitloop equal to its reverse
	// does the circle of bits have an axis of symmetry
	vector<unsigned char> rev = copyArray(n, bs);
	rev = reverse(n, rev);
	bool b = blEqual(n, bs, rev);
	return b;
}


vector<unsigned char> link(int n, vector<unsigned char> bs) {
	// return the link of a given bitloop
	vector<unsigned char> lnk = initArray(n);
	for (int i = 0; i < n; i++) {
		if (getBit(bs, i) ^ getBit(bs, (i + 1) % n)) {
			setBit(lnk, i);
		}
	}
	return lnk;
}


vector<unsigned char> sublink(int n, vector<unsigned char> bs) {
	// return the sublink of a given bitloop
	if (parity(n, bs)) {
		return {};
	}
	vector<unsigned char> slink = initArray(n);
	for (int i = 0; i < n - 1; i++) {
		if (getBit(slink, i) ^ getBit(bs, i)) {
			setBit(slink, i + 1);
		}
	}
	return slink;
}


//vector<unsigned char> pattern(int n, vector<unsigned char> bs);
vector<unsigned char> pattern(int n, vector<unsigned char> bs) {
	// returns the pattern of a given bitloop
	int div = 0; // divisor of n
	bool match = 0;
	while (!match) {
		while (n % ++div != 0) {} // div = next divisor
		match = 1;
		// i is the partition index within the bitstring
		for (int i = 0; match && i < n / div; i++) {
			// j is the bit within partition_i
			for (int j = 0; match && j < div; j++) {
				if (getBit(bs, j) != getBit(bs, div * i + j)) {
					match = 0;
				}
			}
		}
	}
	vector<unsigned char> pat = initArray(div);
	for (int i = 0; i < div; i++) {
		if (getBit(bs, i)) {
			setBit(pat, i);
		}
	}
	return pat;
}


//vector<unsigned char> root(int n, vector<unsigned char> bs);
vector<unsigned char> root(int n, vector<unsigned char> bs) {
	vector<unsigned char> lnk = copyArray(n, bs);
	vector<unsigned char> rt = initArray(n);
	bool match = 0;
	vector<vector<unsigned char>> treecycle;
	while (!match) {
		treecycle.push_back(lnk);
		lnk = link(n, lnk);
		for (int i = 0; !match && i < treecycle.size(); i++) {
			if (blEqual(n, lnk, treecycle[i])) {
				match = 1;
				rt = lnk;
			}
		}
	}
	return rt;
}


vector<vector<unsigned char>> powersetMinValRots(int n) {
	int cardinality = pow2(n);
	vector<vector<unsigned char>> powerset(cardinality, vector<unsigned char>((n + 7) / 8));
	for (int i = 0; i < cardinality; i++) {
		powerset[i] = minValRot(n, decToBs(n, i));
	}
	vector<vector<unsigned char>> powersetMVR;
	powersetMVR.push_back(powerset[0]);
	int index = 0;
	for (int i = 1; i < cardinality; i++) {
		int bs1 = bsToDec(n, powerset[i]);
		int bs2 = bsToDec(n, powersetMVR[index]);
		if (bs1 > bs2) {
			powersetMVR.push_back(powerset[i]);
			index++;
		}
	}
	return powersetMVR;
}


vector<unsigned char> chain(int n, vector<unsigned char> bs, int& cycleOrder) {
	// returns a single bitstring representing the whole chain
	//
	// like root(), but returns the minimum value root in the respective cycle
	// roots (which are bitloops) are each represented by their MVR
	// so it is the MVRs of these roots that are compared
	// 
	// initialize cycleOrder as "int cycleOrder;"
	// and then "...chain(n, bs, &cycleOrder)."
	// *cycleOrder is a pointer to the value declared in chains()
	vector<unsigned char> lnk = bs;
	bool match = 0;
	// treecycle is a series of successive links
	// it ends with the full cycle in the chain
	// it may start in a tree that leads into the cycle
	// that happens when the first bitloop is not a root
	vector<vector<unsigned char>> treecycle;
	while (!match) {
		treecycle.push_back(lnk);
		lnk = link(n, lnk);
		for (int i = 0; !match && i < treecycle.size(); i++) {
			if (blEqual(n, lnk, treecycle[i])) {
				match = 1;
				cycleOrder = treecycle.size() - i;
			}
		}
	}
	// check every node in the cycle, find the lowest value
	int minVal = (1 << n);
	for (int i = 0; i < cycleOrder; i++) {
		int val = bsToDec(n, minValRot(n, lnk));
		if (val < minVal) {
			minVal = val;
		}
		lnk = link(n, lnk);
	}
	return decToBs(n, minVal);
}


vector<vector<unsigned char>> chains(int n) {
	// returns every chain that partitions P_n
	// each chain is represented by the node in its cycle
	// which has the lowest min val rotation of any node in the cycle
	vector<vector<unsigned char>> chs;

	// get list of bitloops
	// each bitloop is represented by the MVR of its rotation class
	vector<vector<unsigned char>> psmvr = powersetMinValRots(n);
	for (int i = 0; i < psmvr.size(); i++) {
	}
	// create another array with bools
	// this indicates if the element has already been worked,
	// that is, its respective chain was already added to chs
	vector<unsigned char> psCheckbox = initArray(pow2(n));
	/*
	* maybe this for loop could be done through multithreading
	* all threads would reference the same 2 arrays
	* - psmvr, psmvrCheckbox
	* its possible some elements would be worked multiple times
	* this becomes more of an issue as the cycle order increases
	*/
	for (int i = 0; i < psmvr.size(); i++) {
		// skip element if it has already been checked
		// this needs to index by decimal value of piomvr_i
		if (getBit(psCheckbox, bsToDec(n, psmvr[i]))) {
			continue;
		}
		// identify respective chain, add to chs
		int cycleOrder;
		vector<unsigned char> ch = chain(n, psmvr[i], cycleOrder);
		chs.push_back(ch);
		// identify all bitloops in the chain (represented by MVR)
		vector<vector<unsigned char>> chNodes = chainNodes(n, ch, cycleOrder);
		// check them all off
		for (int j = 0; j < chNodes.size(); j++) {
			// get index to check off
			// that is the decimal value of the chain node's MVR
			// psmvr and treeNodes store only MVRs, so they may be compared
			int k = bsToDec(n, chNodes[j]);
			setBit(psCheckbox, k);
		}
	}
	// now all the MVRs in the powerset are checked off
	return chs;
}


vector<vector<unsigned char>> chainNodes(int n, vector<unsigned char> ch, int cycleOrder) {
	// returns an array of every node in the chain
	vector<vector<unsigned char>> chNodes;
	vector<unsigned char> lnk = link(n, ch);
	for (int i = 0; i < cycleOrder; i++) {
		vector<vector<unsigned char>> trNodes = treeNodes(n, lnk);
		for (int j = 0; j < trNodes.size(); j++) {
			chNodes.push_back(trNodes[j]);
		}
		lnk = link(n, lnk);
	}
	return chNodes;
}


vector<vector<unsigned char>> treeNodes(int n, vector<unsigned char> rt) {
	// returns an array of every node in the tree
	vector<vector<unsigned char>> trNodes;
	trNodes.push_back(minValRot(n, rt));
	vector<unsigned char> sublnk = sublinkOfRootOffCycle(n, rt);
	makeTree(n, trNodes, sublnk);
	return trNodes;
}


void makeTree(int n, vector<vector<unsigned char>>& tns, vector<unsigned char> sublnk) {
	// treeNodes helper function, recursive
	tns.push_back(minValRot(n, sublnk));
	sublnk = sublink(n, sublnk);
	if (!sublnk.empty()) {
		vector<unsigned char> inv = minValRot(n, inverse(n, sublnk));
		makeTree(n, tns, sublnk);
		if (!blEqual(n, sublnk, inv)) {
			makeTree(n, tns, inv);
		}
	}
}


vector<unsigned char> sublinkOfRootOffCycle(int n, vector<unsigned char> rt) {
	// the sublink of a given root that is not also a root
	// first take the root's two sublinks
	// then for each sublink, take successive sublinks until
	// sublink() returns {}, i.e. result.empty() == true
	// whichever leads to {} first is the off-cycle sublink
	vector<unsigned char> treeNode1 = sublink(n, rt);
	vector<unsigned char> treeNode2 = inverse(n, treeNode1);
	while (!treeNode1.empty() && !treeNode2.empty()) {
		treeNode1 = sublink(n, treeNode1);
		treeNode2 = sublink(n, treeNode2);
	}
	return treeNode1.empty() ? sublink(n, rt) : inverse(n, sublink(n, rt));
}
























































